targets/TARGET_NORDIC/TARGET_NRF5x/TARGET_SDK_14_2/libraries/atomic_fifo/nrf_atfifo.c

```diff
@@ -0,0 +1,160 @@
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+ * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ * 
+ */
+
+#include <string.h>
+#include "app_util.h"
+#include "nrf_atfifo.h"
+#include "nrf_atfifo_internal.h"
+
+
+/* Unions testing */
+STATIC_ASSERT(sizeof(nrf_atfifo_postag_t) == sizeof(uint32_t));
+
+
+ret_code_t nrf_atfifo_init(nrf_atfifo_t * const p_fifo, void * p_buf, uint16_t buf_size, uint16_t item_size)
+{
+    if (NULL == p_buf)
+    {
+        return NRF_ERROR_NULL;
+    }
+    if (0 != (buf_size % item_size))
+    {
+        return NRF_ERROR_INVALID_LENGTH;
+    }
+
+    p_fifo->p_buf     = p_buf;
+    p_fifo->tail.tag  = 0;
+    p_fifo->head.tag  = 0;
+    p_fifo->buf_size  = buf_size;
+    p_fifo->item_size = item_size;
+
+    return NRF_SUCCESS;
+}
+
+
+ret_code_t nrf_atfifo_clear(nrf_atfifo_t * const p_fifo)
+{
+    bool released = nrf_atfifo_space_clear(p_fifo);
+    return released ? NRF_SUCCESS : NRF_ERROR_BUSY;
+}
+
+
+ret_code_t nrf_atfifo_alloc_put(nrf_atfifo_t * const p_fifo, void const * p_var, size_t size, bool * const p_visible)
+{
+    nrf_atfifo_item_put_t context;
+    bool visible;
+    void * p_data = nrf_atfifo_item_alloc(p_fifo, &context);
+    if (NULL == p_data)
+    {
+        return NRF_ERROR_NO_MEM;
+    }
+
+    memcpy(p_data, p_var, size);
+
+    visible = nrf_atfifo_item_put(p_fifo, &context);
+    if (NULL != p_visible)
+    {
+        *p_visible = visible;
+    }
+    return NRF_SUCCESS;
+}
+
+
+void * nrf_atfifo_item_alloc(nrf_atfifo_t * const p_fifo, nrf_atfifo_item_put_t * p_context)
+{
+    if (nrf_atfifo_wspace_req(p_fifo, &(p_context->last_tail)))
+    {
+        return ((uint8_t*)(p_fifo->p_buf)) + p_context->last_tail.pos.wr;
+    }
+    return NULL;
+}
+
+
+bool nrf_atfifo_item_put(nrf_atfifo_t * const p_fifo, nrf_atfifo_item_put_t * p_context)
+{
+    if ((p_context->last_tail.pos.wr) == (p_context->last_tail.pos.rd))
+    {
+        nrf_atfifo_wspace_close(p_fifo);
+        return true;
+    }
+    return false;
+}
+
+
+ret_code_t nrf_atfifo_get_free(nrf_atfifo_t * const p_fifo, void * const p_var, size_t size, bool * p_released)
+{
+    nrf_atfifo_item_get_t context;
+    bool released;
+    void const * p_s = nrf_atfifo_item_get(p_fifo, &context);
+    if (NULL == p_s)
+    {
+        return NRF_ERROR_NOT_FOUND;
+    }
+
+    memcpy(p_var, p_s, size);
+
+    released = nrf_atfifo_item_free(p_fifo, &context);
+    if (NULL != p_released)
+    {
+        *p_released = released;
+    }
+    return NRF_SUCCESS;
+}
+
+
+void * nrf_atfifo_item_get(nrf_atfifo_t * const p_fifo, nrf_atfifo_item_get_t * p_context)
+{
+    if (nrf_atfifo_rspace_req(p_fifo, &(p_context->last_head)))
+    {
+        return ((uint8_t*)(p_fifo->p_buf)) + p_context->last_head.pos.rd;
+    }
+    return NULL;
+}
+
+
+bool nrf_atfifo_item_free(nrf_atfifo_t * const p_fifo, nrf_atfifo_item_get_t * p_context)
+{
```

Comment:
need to check p_context !=NULL

Reply:
Note that this file is from 3rd party, should be sent upstream rather than changing here